File: ADT/dfs_bfs/433.js

var minMutation = function (start, end, bank) {
    const bankset = new Set(bank)
    const visited = new Set()
    const q = []
    q.push(start)
    visited.add(start)
    let level = 0
    while (q.length > 0) {
        let n = q.length
        while (n--) {
            const curr = q.shift()
            if (curr === end)
                return level
            for (const ch of "ACGT") {
                for (let i = 0; i < curr.length; i++) {
                    const neighbor = curr.slice(0, i) + ch + curr.slice(i + 1)
                    if (!visited.has(neighbor) && bankset.has(neighbor)) {
                        visited.add(neighbor)
                        q.push(neighbor)
                    }
                }
            }
        }
        level++
    }
    return -1
};

File: ADT/dfs_bfs/433.py
from collections import deque


class Solution:
    def minMutation(self, start: str, end: str, bank: list[str]) -> int:
        bankset = set(bank)
        visited = set()
        que = deque([start])
        visited.add(start)
        
        level = 0
        
        while que:
            n = len(que)
            
            for _ in range(n):
                curr = que.popleft()
                
                if curr == end:
                    return level
                
                for ch in "ACGT":
                    for i in range(len(curr)):
                        neighbour = curr[:i] + ch + curr[i+1:]
                        
                        if neighbour not in visited and neighbour in bankset:
                            visited.add(neighbour)
                            que.append(neighbour)
            
            level += 1
        
        return -1